compensations/src/main/java/org/jboss/narayana/compensations/internal/ParticipantImpl.java

```diff
@@ -138,38 +153,79 @@ public void error() throws SystemException {
 
     }
 
-    /**
-     * Increments the counter of the Compensations participants in the transaction.
-     */
-    private void incrementParticipantsCounter() {
+    @Override
+    public boolean saveState(OutputObjectState state) {
+        LOGGER.tracef("Persisting state: '%s'", this);
+
+        try {
+            state.packString(currentTransactionId);
+            state.packString(participantId);
+            packHandler(state, compensationHandler);
+            packHandler(state, confirmationHandler);
+        } catch (IOException e) {
+            LOGGER.warnf(e, "Failed to persist state");
+            return false;
+        }
 
-        synchronized (PARTICIPANT_COUNTERS) {
-            final AtomicInteger counter = PARTICIPANT_COUNTERS.get(currentTX);
+        return true;
+    }
 
-            if (counter == null) {
-                PARTICIPANT_COUNTERS.put(currentTX, new AtomicInteger(1));
-            } else {
-                counter.incrementAndGet();
-            }
+    @Override
+    public boolean restoreState(InputObjectState state) {
+        try {
+            currentTransactionId = state.unpackString();
+            participantId = state.unpackString();
+            compensationHandler = unpackHandler(state, CompensationHandler.class);
+            confirmationHandler = unpackHandler(state, ConfirmationHandler.class);
+        } catch (IOException e) {
+            LOGGER.warnf(e, "Failed to restore state");
+            return false;
         }
+
+        LOGGER.tracef("Restored state: '%s'", this);
+
+        return true;
     }
 
-    /**
-     * Decrements the counter of the Compensations participants in the transaction.
-     * CompensationContext of the current transaction is destroyed once the counter reaches 0.
-     */
-    private void decrementParticipantsCounter() {
+    @Override
+    public String toString() {
+        return "ParticipantImpl{currentTransactionId='" + currentTransactionId + "', compensationHandler=" + compensationHandler
+                + ", confirmationHandler=" + confirmationHandler + ", transactionLoggedHandler=" + transactionLoggedHandler
+                + ", applicationClassloader=" + applicationClassloader + "}";
+    }
+
+    private <T> boolean isHandlerPersistable(T handler) {
+        return handler != null && handler instanceof Serializable;
+    }
+
+    private <T> void packHandler(OutputObjectState state, T handler) throws IOException {
+        if (!isHandlerPersistable(handler)) {
+            state.packBoolean(false);
```

Comment:
Log a warning if the handler was not serializable - add a note to the docs to specify that is the expectation